docs(deploy/GKE): add guide to deploy toolbox to GKE (#190)
googleapis/genai-toolbox: docs/en/how-to/deploy_gke.md

---

> title: "Deploy to GKE"

Reviewer: ```suggestion
title: "Deploy to Kubernetes"
```

type: docs
weight: 2
description: >

>   How to set up and configure Toolbox to deploy on Google Kubernetes Engine (GKE).

Reviewer: ```suggestion
  How to set up and configure Toolbox to deploy on Kubernetes with Google Kubernetes Engine (GKE).
```

---

## Before you begin

> 1. [Install the `gcloud` CLI](https://cloud.google.com/sdk/docs/install).
> 
> 1. Set the PROJECT_ID environment variable:
> 
>     ```bash
>     export PROJECT_ID="my-project-id"
>     ```

Reviewer: super-nit: 
```suggestion

1. Set the PROJECT_ID environment variable:

    ```bash
    export PROJECT_ID="my-project-id"
    ```

1. [Install the `gcloud` CLI](https://cloud.google.com/sdk/docs/install).
```

   by GKE. Verify if you have `kubectl` installed:

    ```bash
    kubectl version --client
    ```
   

1. If needed, install `kubectl` component using the Google Cloud CLI:

   ```bash
   gcloud components install kubectl
   ```

## Create a service account

1. Specify a name for your service account with an environment variable:

    ```bash
    export SA_NAME=toolbox
    ```

1. Create a backend service account:

    ```bash
    gcloud iam service-accounts create $SA_NAME
    ```

1.  Grant any IAM roles necessary to the IAM service account. Each source have a
    list of necessary IAM permissions listed on it's page. The example below is
    for cloud sql postgres source:

    ```bash
    gcloud projects add-iam-policy-binding $PROJECT_ID \
        --member serviceAccount:$SA_NAME@$PROJECT_ID.iam.gserviceaccount.com \
        --role roles/cloudsql.client
    ```

    - [AlloyDB IAM Identity](../resources/sources/alloydb-pg.md#iam-permissions)
    - [CloudSQL IAM Identity](../resources/sources/cloud-sql-pg.md#iam-permissions)
    - [Spanner IAM Identity](../resources/sources/spanner.md#iam-permissions)

> ## Deploying to GKE

Reviewer: ```suggestion
## Deploy to Kubernetes 
```

1. Set environment variables:

    ```bash
    export cluster_name=toolbox-cluster
    export deployment_name=toolbox
    export service_name=toolbox-service
    export region=us-central1
    export namespace=toolbox-namespace
    export secret_name=toolbox-config

>     export ksa_name=toolbox-service-account

Reviewer: nit: Would CAPITALIZE these

   configures `kubectl` to use the cluster.

    ```bash
    gcloud container clusters get-credentials $cluster_name \
        --region=$region \
        --project=$PROJECT_ID
    ```

1. View the current context for `kubectl`.

    ```bash
    kubectl config current-context
    ```

1. Create namespace for the deployment.

    ```bash
    kubectl create namespace $namespace
    ```

1. Create a Kubernetes Service Account (KSA).

    ```bash
    kubectl create serviceaccount $ksa_name --namespace $namespace
    ```

1. Enable the IAM binding between Google Service Account (GSA) and Kubernetes
   Service Account (KSA).

    ```bash
    gcloud iam service-accounts add-iam-policy-binding \
        --role="roles/iam.workloadIdentityUser" \
        --member="serviceAccount:$PROJECT_ID.svc.id.goog[$namespace/$ksa_name]" \
        $SA_NAME@$PROJECT_ID.iam.gserviceaccount.com
    ```

1. Add annotation to KSA to complete binding:

    ```bash
    kubectl annotate serviceaccount \
        $ksa_name \
        iam.gke.io/gcp-service-account=$SA_NAME@$PROJECT_ID.iam.gserviceaccount.com \
        --namespace $namespace
    ```

> 1. Prepare the kubernetes Secrets (`tools.yaml` file).

Reviewer: ```suggestion
1. Prepare the Kubernetes secret for your `tools.yaml` file.
```
